### PR TITLE
expression: fix errors set utc_timestamp precision | tidb-test=pr/2406

### LINKED ISSUE #56451: UTC_TIMESTAMP(7) gives different error code and message from MySQL
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
SELECT UTC_TIMESTAMP(7);
```

### 2. What did you expect to see? (Required)
MySQL 8.0.32:
```sql
mysql> SELECT UTC_TIMESTAMP(7);
ERROR 1426 (42000): Too-big precision 7 specified for 'utc_timestamp'. Maximum is 6.
```
### 3. What did you see instead (Required)
TiDB:
```sql
tidb> SELECT UTC_TIMESTAMP(7);
ERROR 1105 (HY000): Too-big precision 7 specified for 'curtime'. Maximum is 6
```
Which means that:
- error code is different
- sqlstate is different
- function name is not correct.
### 4. What is your TiDB version? (Required)

```
tidb_version(): Release Version: v8.3.0
Edition: Community
Git Commit Hash: 1a0c3ac3292fff7742faa0c00a662ccb66ba40db
Git Branch: HEAD
UTC Build Time: 2024-08-20 10:23:00
GoVersion: go1.21.10
Race Enabled: false
Check Table Before Drop: false
Store: tikv
```

### PR DESCRIPTION
**Pull request by chagelo**
### What problem does this PR solve?

Issue Number: close #56451

Problem Summary:

### What changed and how does it work?

Correct the error when set invalid precision for `utc_timestamp`

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix errors when set invalid precision of date or time related functions, such as too big precision.
```

### CODE REVIEW

**Review by mjonss — COMMENTED** [review]
Looks good, just missing the statement in the .test file.

**Comment by chagelo** [comment]
Thank you very much for the suggestions!

**Review by mjonss — APPROVED** [review]
Looks good to me!

I have re-recorded the test results for the failing CI test `idc-jenkins-ci-tidb/mysql-test`, which is a private repository for various integration tests and mysql compatibility tests. so hopefully the tests will pass.

**Comment by mjonss** [comment]
@chagelo I will look for reviewers that have code ownership for expression and parser, so this can be fully approved and merged, but it may take a few days, due to weekend and holidays.

Thank you again for your contribution!

**Comment by chagelo** [comment]
Thank you as well! 😁

File: pkg/expression/builtin_time_vec.go [hunk 1]
```diff
@@ -547,7 +547,7 @@ func (b *builtinNowWithArgSig) vecEvalTime(ctx EvalContext, input *chunk.Chunk,
 		fsp := 0
 		if !bufFsp.IsNull(i) {
 			if fsps[i] > int64(types.MaxFsp) {
-				return errors.Errorf("Too-big precision %v specified for 'now'. Maximum is %v", fsps[i], types.MaxFsp)
```
Comment (chagelo) [inline pkg/expression/builtin_time_vec.go:550]: There is a typos error of mine, so I change it from `fsp` to `fsps[i]` and then merge it with previos commit.

File: pkg/expression/builtin_time_vec.go [hunk 2]
```diff
@@ -547,7 +547,7 @@ func (b *builtinNowWithArgSig) vecEvalTime(ctx EvalContext, input *chunk.Chunk,
...
+				return types.ErrTooBigPrecision.GenWithStackByArgs(fsps[i], "now", types.MaxFsp)
```
Comment (chagelo) [inline pkg/expression/builtin_time_vec.go:550]: There is a typos error of mine, so I change it from `fsp` to `fsps[i]` and then merge it with previos commit.

**Review by dveeden — APPROVED** [review]
Why `ref ...` and not `closes ...` ? What part of the issue isn't fixed by this PR?

I think this should have a release note.

**Comment by dveeden** [comment]
There are a few places that check for a `fsp` that's less than `MinFsp`. But looks like there is no way to reach that error as the parser is not allowing this. 

```
        } else if fsp > int64(types.MaxFsp) {
                return types.ZeroTime, true, types.ErrTooBigPrecision.GenWithStackByArgs(fsp, "now", types.MaxFsp)
        } else if fsp < int64(types.MinFsp) {
                return types.ZeroTime, true, errors.Errorf("Invalid negative %d specified, must in [0, 6]", fsp)
        }
```

Maybe:
- Remove the check for `fsp < MinFsp`
- Add a comment to indicate that this should never be reached and/or change to message to indicate this
- Add tests for things like `NOW(-1)` to ensure it gets rejected by the parser

Side note: Looks like there is a MySQL bug in the error message that we don't follow:

MySQL 9.0.1:
```
sql> SELECT NOW(99999);
ERROR: 1426 (42000): Too-big precision 159 specified for 'now'. Maximum is 6.
```

TiDB:
```
sql> SELECT NOW(99999);
ERROR: 1426 (42000): Too-big precision 99999 specified for 'now'. Maximum is 6.
```

**Comment by dveeden** [comment]
For the bug in MySQL:
https://bugs.mysql.com/bug.php?id=116307

**Comment by dveeden** [comment]
Inspired by the MySQL bug it looks like we have a similar bug (but way less severe). This also answers my question about negative fsp's somewhat:
```
sql> SELECT SYSDATE(999999999999999999);
ERROR: 1426 (42000): Too-big precision 999999999999999999 specified for 'sysdate'. Maximum is 6.

sql> SELECT SYSDATE(9999999999999999999);
ERROR: 1105 (HY000): Invalid negative -8446744073709551617 specified, must in [0, 6]
```

**Comment by chagelo** [comment]
@dveeden Could you show me the result  `SELECT NOW(-1);` and `SELECT NOW(256);` of MySQL? I have not installed MySQL 9.0.1 Version.

**Comment by chagelo** [comment]
1. For MySQL, it uses the last byte in binary representation as the precision, so `SELECT NOW(256)` equals to `SELECT NOW(1)`.
2. And if precision has more than 32 bits like 2147483647 + 1, it return a SQL syntax error.
3. I'll add a comment.

**Comment by dveeden** [comment]
MariaDB seems to be more strict than MySQL here. Looks like MariaDB accepts larger numbers in the parser but checks them later on.
```
mysql-11.4.2-MariaDB-ubu2404> SELECT NOW(9999999999999999999);
ERROR 1426 (42000): Too big precision specified for 'current_timestamp'. Maximum is 6
mysql-11.4.2-MariaDB-ubu2404> SELECT NOW(99999999999999999999);
ERROR 1064 (42000): Only integers allowed as number here near '99999999999999999999)' at line 1
```

File: pkg/expression/builtin_time.go [hunk 3]
```diff
@@ -2429,14 +2429,14 @@ func (b *builtinUTCTimestampWithArgSig) evalTime(ctx EvalContext, row chunk.Row)
...
-	if !isNull && num > int64(types.MaxFsp) {
-		return types.ZeroTime, true, errors.Errorf("Too-big precision %v specified for 'utc_timestamp'. Maximum is %v", num, types.MaxFsp)
-	}
-	if !isNull && num < int64(types.MinFsp) {
-		return types.ZeroTime, true, errors.Errorf("Invalid negative %d specified, must in [0, 6]", num)
+	// Here fsp must be greater than or equal to zero, there is a bug of parser for negetive precision now.
+	// There is a bug of MySQL, but not fixed now, so we keep it be compatible with MySQL. (#56453)
+	fspu8 := uint8(num)
+	if !isNull && fspu8 > uint8(types.MaxFsp) {
+		return types.ZeroTime, true, types.ErrTooBigPrecision.GenWithStackByArgs(fspu8, "utc_timestamp", types.MaxFsp)
```
[the remark this answers is not in the document]
Comment (chagelo) [reply pkg/expression/builtin_time.go:2436]: Could we just open this pr? After MySQL fix it, then I'll fix this.

**Comment by dveeden** [comment]
@chagelo are the tests failing for something related to this PR?

**Comment by chagelo** [comment]
@dveeden yes, it is.
1. Because MySQL has the bug, I don't know what should I do to make it be compatible with MySQL.
2. I read the parser code. I want parser to limit precision's type like what MySQL does, but I don't understand the `yyParseTab`, it seems complicated.

**Comment by dveeden** [comment]
> @dveeden yes, it is.
> 
>     1. Because MySQL has the bug, I don't know what should I do to make it be compatible with MySQL.

I don't think we should follow the behavior of https://bugs.mysql.com/bug.php?id=116307 as I don't think people are relying on this behavior or should rely on this behavior.

>     2. I read the parser code. I want parser to limit precision's type like what MySQL does, but I don't understand the `yyParseTab`, it seems complicated.

`yyParseTab` is in `pkg/parser/parser.go` which is generated by `goyacc` based on `pkg/parser/parser.y`, so you need to ecit `parser.y` and run `make parser`.
 
>     3. It seems not good to throw a syntax error in `pkg/expression/builtin_time.go`(like negative or too long number precision), it should done by parser, it's right? I'm not sure about this.

I'm not sure. One of the issues with having the parser deal with things is that it might not have the right context to fill in the fields of the error message (e.g. the name of the table or the name of the function).

**Comment by dveeden** [comment]
From the issue description:

- error code is different
- sqlstate is different
- function name is not correct.

It looks like these are all fixed. 

The following might be seen as a separate bug if it is not a regression:
```
mysql-8.0.11-TiDB-v8.4.0-alpha-326-ga2b0d7b115> SELECT UTC_TIMESTAMP(9990);
+----------------------------+
| UTC_TIMESTAMP(9990)        |
+----------------------------+
| 2024-10-14 07:12:01.054447 |
+----------------------------+
1 row in set (0.00 sec)
```

Note that this is somewhat similar to https://github.com/pingcap/tidb/issues/46372#issuecomment-2405559661

**Comment by chagelo** [comment]
```
SELECT CURRENT_TIME(-1);
Error 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 21 near "-1);" 
SELECT CURRENT_TIME(2147483648);
Error 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use
```

It seems TiDB parser handles the negative precision, but not resolve the overlong number precision, parser should have dealt with this, because this error contains some context.

**Comment by chagelo** [comment]
@dveeden Required review. Thanks.

**Review by dveeden — APPROVED** [review]

File: pkg/expression/builtin_time_vec.go [hunk 4]
```diff
@@ -404,12 +404,12 @@ func (b *builtinUTCTimeWithArgSig) vecEvalDuration(ctx EvalContext, input *chunk
 		if result.IsNull(i) {
 			continue
 		}
+		if i64s[i] > int64(math.MaxInt32) || i64s[i] < int64(types.MinFsp) {
```
Comment (XuHuaiyu) [inline pkg/expression/builtin_time_vec.go:407]:
```suggestion
		fsp := i64s[i]
		if fsp > int64(math.MaxInt32) || fsp < int64(types.MinFsp) {
```

File: pkg/expression/builtin_time_vec.go [hunk 5]
```diff
@@ -1407,12 +1407,12 @@ func (b *builtinUTCTimestampWithArgSig) vecEvalTime(ctx EvalContext, input *chun
...
```
Comment (XuHuaiyu) [inline pkg/expression/builtin_time_vec.go:1410]: ditto

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**